File: backend/app/services/market.py
"""Market data access, backed by Yahoo Finance (via yfinance).

Quotes are cached in-process for a short TTL so leaderboard/portfolio
valuations don't hammer the API.
"""

import logging
import time
from datetime import date, timedelta
from decimal import Decimal

import yfinance as yf

from app.config import settings

logger = logging.getLogger(__name__)

_quote_cache: dict[str, tuple[Decimal, float]] = {}
_range_cache: dict[tuple[str, date, date], tuple[list, float]] = {}
RANGE_CACHE_TTL = 3600


class UnknownTickerError(Exception):
    def __init__(self, ticker: str):
        self.ticker = ticker
        super().__init__(f"No price data found for ticker '{ticker}'")


def get_quote(ticker: str) -> Decimal:
    """Latest price for a ticker, cached for `quote_cache_ttl_seconds`."""
    ticker = ticker.upper().strip()
    cached = _quote_cache.get(ticker)
    if cached and time.time() - cached[1] < settings.quote_cache_ttl_seconds:
        return cached[0]

    price = _fetch_price(ticker)
    if price is None:
        raise UnknownTickerError(ticker)

    _quote_cache[ticker] = (price, time.time())
    return price


def _fetch_price(ticker: str) -> Decimal | None:
    yf_ticker = yf.Ticker(ticker)
    try:
        last = yf_ticker.fast_info.last_price
        if last:
            return Decimal(str(round(float(last), 4)))
    except Exception:
        logger.debug("fast_info failed for %s, falling back to history", ticker)

    try:
        hist = yf_ticker.history(period="5d")
        if not hist.empty:
            return Decimal(str(round(float(hist["Close"].iloc[-1]), 4)))
    except Exception:
        logger.warning("history lookup failed for %s", ticker, exc_info=True)
    return None


def search(query: str, limit: int = 10) -> list[dict]:
    """Search tickers by name or symbol."""
    try:
        results = yf.Search(query, max_results=limit).quotes
    except Exception:
        logger.warning("ticker search failed for %r", query, exc_info=True)
        return []

    return [
        {
            "ticker": item.get("symbol"),
            "name": item.get("shortname") or item.get("longname") or "",
            "exchange": item.get("exchange") or "",
            "type": item.get("quoteType") or "",
        }
        for item in results
        if item.get("symbol")
    ]


def get_history_range(ticker: str, start: date, end: date) -> list[tuple[date, Decimal]]:
    """Daily closes between two dates (inclusive), oldest first. Cached for an
    hour since historical closes don't change."""
    ticker = ticker.upper().strip()
    key = (ticker, start, end)
    cached = _range_cache.get(key)
    if cached and time.time() - cached[1] < RANGE_CACHE_TTL:
        return cached[0]

    try:
        hist = yf.Ticker(ticker).history(
            start=start.isoformat(), end=(end + timedelta(days=1)).isoformat()
        )
    except Exception:
        logger.warning("historical range fetch failed for %s", ticker, exc_info=True)
        return []
    rows = [
        (index.date(), Decimal(str(round(float(row["Close"]), 4))))
        for index, row in hist.iterrows()
    ]
    _range_cache[key] = (rows, time.time())
    return rows


def get_price_on(ticker: str, on_date: date) -> Decimal | None:
    """Closing price on a given date, falling back to the most recent close
    before it (weekends, holidays). None if the ticker has no data by then."""
    rows = get_history_range(ticker, on_date - timedelta(days=14), on_date)
    rows = [row for row in rows if row[0] <= on_date]
    return rows[-1][1] if rows else None


def get_history(ticker: str, period: str = "1mo") -> list[dict]:
    """Daily closes for a ticker over a yfinance period string (1mo, 3mo, 1y...)."""
    hist = yf.Ticker(ticker.upper().strip()).history(period=period)
    if hist.empty:
        raise UnknownTickerError(ticker)
    return [
        {"date": index.date().isoformat(), "close": round(float(row["Close"]), 4)}
        for index, row in hist.iterrows()
    ]

File: backend/src/market.js
/** Market data via Yahoo Finance's public chart/search endpoints (the same
 * ones yfinance wraps). In-isolate caches keep request volume low. */

import { r4 } from "./util.js";

const HEADERS = { "User-Agent": "Mozilla/5.0 (compatible; FakeStockGame/1.0)" };
const QUOTE_TTL_MS = 5 * 60 * 1000;
const RANGE_TTL_MS = 60 * 60 * 1000;

const quoteCache = new Map(); // ticker -> {price, at}
const rangeCache = new Map(); // key -> {rows, at}

export class UnknownTickerError extends Error {
  constructor(ticker) {
    super(`No price data found for ticker '${ticker}'`);
    this.ticker = ticker;
  }
}

async function fetchChart(ticker, query) {
  const url = `https://query1.finance.yahoo.com/v8/finance/chart/${encodeURIComponent(ticker)}?${query}`;
  const response = await fetch(url, { headers: HEADERS });
  if (!response.ok) return null;
  const body = await response.json();
  return body?.chart?.result?.[0] ?? null;
}

function chartRows(result) {
  const timestamps = result?.timestamp ?? [];
  const adjusted =
    result?.indicators?.adjclose?.[0]?.adjclose ??
    result?.indicators?.quote?.[0]?.close ??
    [];
  const rows = [];
  for (let i = 0; i < timestamps.length; i++) {
    if (adjusted[i] !== null && adjusted[i] !== undefined) {
      rows.push({
        date: new Date(timestamps[i] * 1000).toISOString().slice(0, 10),
        close: r4(adjusted[i]),
      });
    }
  }
  return rows;
}

export async function getQuote(ticker) {
  ticker = ticker.toUpperCase().trim();
  const cached = quoteCache.get(ticker);
  if (cached && Date.now() - cached.at < QUOTE_TTL_MS) return cached.price;

  const result = await fetchChart(ticker, "range=5d&interval=1d");
  const rows = result ? chartRows(result) : [];
  const price = result?.meta?.regularMarketPrice ?? rows.at(-1)?.close ?? null;
  if (price === null || price === undefined) throw new UnknownTickerError(ticker);

  const rounded = r4(price);
  quoteCache.set(ticker, { price: rounded, at: Date.now() });
  return rounded;
}

/** Daily (adjusted) closes between two ISO dates inclusive, oldest first. */
export async function getHistoryRange(ticker, startISO, endISO) {
  ticker = ticker.toUpperCase().trim();
  const key = `${ticker}:${startISO}:${endISO}`;
  const cached = rangeCache.get(key);
  if (cached && Date.now() - cached.at < RANGE_TTL_MS) return cached.rows;

  const period1 = Math.floor(Date.parse(`${startISO}T00:00:00Z`) / 1000);
  const period2 = Math.floor(Date.parse(`${endISO}T00:00:00Z`) / 1000) + 86400;
  const result = await fetchChart(
    ticker, `period1=${period1}&period2=${period2}&interval=1d&events=div%2Csplit`
  );
  const rows = result ? chartRows(result) : [];
  rangeCache.set(key, { rows, at: Date.now() });
  return rows;
}

/** Close on a given date, falling back to the most recent close before it
 * (weekends/holidays). Null if the ticker has no data by then. */
export async function getPriceOn(ticker, isoDate) {
  const start = new Date(`${isoDate}T00:00:00Z`);
  start.setUTCDate(start.getUTCDate() - 14);
  const rows = await getHistoryRange(ticker, start.toISOString().slice(0, 10), isoDate);
  const eligible = rows.filter((row) => row.date <= isoDate);
  return eligible.at(-1)?.close ?? null;
}

export async function searchStocks(query, limit = 10) {
  const url = `https://query1.finance.yahoo.com/v1/finance/search?q=${encodeURIComponent(query)}&quotesCount=${limit}&newsCount=0`;
  try {
    const response = await fetch(url, { headers: HEADERS });
    if (!response.ok) return [];
    const body = await response.json();
    return (body?.quotes ?? [])
      .filter((item) => item.symbol)
      .map((item) => ({
        ticker: item.symbol,
        name: item.shortname || item.longname || "",
        exchange: item.exchange || "",
        type: item.quoteType || "",
      }));
  } catch {
    return [];
  }
}

export async function getHistory(ticker, period) {
  const result = await fetchChart(ticker.toUpperCase().trim(), `range=${period}&interval=1d`);
  const rows = result ? chartRows(result) : [];
  if (rows.length === 0) throw new UnknownTickerError(ticker);
  return rows;
}
